ledger/participant-integration-api/src/main/scala/platform/usermanagement/CachedUserManagementStore.scala

```diff
@@ -0,0 +1,145 @@
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object CachedUserManagementStore {
+
+  // TODO participant user management: copied from InMemory version
+  case class UserInfo(user: User, rights: Set[UserRight]) {
+    def toStateEntry: (Ref.UserId, UserInfo) = user.id -> this
+  }
+
+  val ExpiryAfterWriteInSeconds: Int = 10
+}
+
+class CachedUserManagementStore(
+    private val delegate: UserManagementStore,
+    expiryAfterWriteInSeconds: Int,
+)(implicit val executionContext: ExecutionContext)
+    extends UserManagementStore {
+
+  private val usersCache: ConcurrentCache[UserId, User] = CaffeineCache[Ref.UserId, User](
+    builder = Caffeine
+      .newBuilder()
+      .expireAfterWrite(Duration.ofSeconds(expiryAfterWriteInSeconds.toLong))
+      // TODO participant user management: Check the choice of the maximum size
+      .maximumSize(10000),
+    // TODO participant user management: Use metrics
+    metrics = None,
+  )
+
+  import com.github.benmanes.caffeine.cache.Caffeine
+
+  private val userRightsCache: ConcurrentCache[UserId, Set[UserRight]] =
+    CaffeineCache[Ref.UserId, Set[UserRight]](
+      builder = Caffeine
+        .newBuilder()
+        .expireAfterWrite(Duration.ofSeconds(expiryAfterWriteInSeconds.toLong))
+        // TODO participant user management: Check the choice of the maximum size
+        .maximumSize(10000),
+      // TODO participant user management: Use metrics
+      metrics = None,
+    )
+
+  // TODO copied from Persistent version
+  private def tapSuccess[T](f: T => Unit)(r: Result[T]): Result[T] = {
+    r match {
+      case Right(v) => f(v)
+      case Left(error) =>
+        error match {
+          case UserNotFound(userId) =>
+            // TODO participant user management: We don't need to invalidate here as deleteUser() already handles it
+            usersCache.invalidate(userId)
+            userRightsCache.invalidate(userId)
+          case _: UserExists =>
+          case _: TooManyUserRights =>
+        }
+    }
+    r
+  }
+
+  override def createUser(user: domain.User, rights: Set[domain.UserRight]): Future[Result[Unit]] =
+    delegate.createUser(user, rights)
+
+  override def getUser(id: UserId): Future[Result[domain.User]] = {
+    usersCache
+      .getIfPresent(id)
+      .fold(
+        delegate
+          .getUser(id)
+          .map(tapSuccess(user => usersCache.put(id, user)))
+      )((user) => Future.successful(Right(user)))
+  }
+
+  override def deleteUser(id: UserId): Future[Result[Unit]] = {
+    delegate
+      .deleteUser(id)
+      .map(tapSuccess { _ =>
+        usersCache.invalidate(id)
+        userRightsCache.invalidate(id)
+      })
+  }
+
+  override def grantRights(
+      id: UserId,
+      rights: Set[domain.UserRight],
+  ): Future[Result[Set[domain.UserRight]]] = {
+    delegate
+      .grantRights(id, rights)
+      .map(
+        tapSuccess(granted =>
+          userRightsCache
+            .getIfPresent(id)
```

Comment:
Cache is not updated in case there are no previous rights in the cache

Reply:
I think I can defend current implementation:
1. If user rights cache entry is missing then we haven't read user rights in a while and we can guess we won't be needing them anytime soon. We will populate cache only on read access (e.g. getUserRights).
2. To populate user rights cache entry we'd have to make an extra db call to read all existing user rights.

Reply:
Out-of-date new approach: eager invalidation on all writes